### PR TITLE
pkg/lwip: allow different types of netdevs as lwIP netifs

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides a change that allows `netdev` of different type to be used as `netifs in lwIP.

So far only one `netdev` driver could be used as lwIP `netif` at the same time.  The number of lwIP `netif` `LWIP_NETIF_NUMOF` then corresponded to the number of `netdev` of only one type. The different types of `netdev` could only be used exclusively. The `netif_add` function was called only for the first used `netdev` driver.

With this PR, `netdev` of different types can be used at the same time. For this purpose, this PR defines a number of `netdev` for each different type `LWIP_NETIF_NUMOF_<netdev>`. The number of lwip `netif` `LWIP_NETIF_NUMOF` is then the sum of all numbers of the different `netdev` `LWIP_NETIF_NUMOF_<netdev>`. The `netif_add` function is called for each used `netdev` type. The index of the according lwIP `netif` is counted up. The default lwIP `netif` can be defined using `LWIP_NETIF_DEFAULT` which is 0 by default.

This PR also fixes the output of the lwip `netif.name` in `printf` of the `ifconfig` command.

The PR was tested with
- `esp32-olimex-evb` with the WiFi and the Ethernet Interface in dual stack mode
   ```
   ET_01:  inet6 fe80::32ae:a4ff:fe47:4ccc
           inet6 2003:c1:e748:6b54:32ae:a4ff:fe47:4ccc
           inet 192.168.2.155
   
   ET_00:  inet6 fe80::32ae:a4ff:fe47:4ccf
           inet6 2003:c1:e748:6b54:32ae:a4ff:fe47:4ccf
           inet 192.168.2.113
   ```
- `esp32-mh-et-live-minikit` with the WiFi interface and a MRF24J40 6LoWPAN interface in IPv6 mode
   ```
   ET_01:  inet6 fe80::260a:c4ff:fe06:a164
           inet6 2003:c1:e748:6b54:260a:c4ff:fe06:a164
   
   L6_00:  inet6 fe80::3e47:8225:e729:73e
   ```

This PR contains commit https://github.com/RIOT-OS/RIOT/pull/14374/commits/b64ce62a8f409726aba651c6ae634011cb2195bb from PR #14374 which can be dropped once PR #14374 is merged.

### Testing procedure

t.b.d.

### Issues/PRs references

Depends on PR #14374.

### CODE REVIEW

**Comment by aabadie** [comment]
#14374 is merged, so please rebase!

File: pkg/lwip/contrib/lwip.c [hunk 1]
```diff
@@ -179,94 +230,106 @@ void lwip_bootstrap(void)
 {
     (void)_netif_add;   /* in case it is not used */
     /* TODO: do for every eligible netdev */
-#ifdef LWIP_NETIF_NUMOF
+#ifdef LWIP_NETIF_USED
+    int _netif = 0;
```
Comment (miri64) [inline pkg/lwip/contrib/lwip.c:234]: Why the `_` prefix? It usually is used to denote private scope, but for a function variable basically all is private (to the function), so usually you use it for global variables and functions only.

File: pkg/lwip/contrib/lwip.c [hunk 2]
```diff
@@ -179,94 +230,106 @@ void lwip_bootstrap(void)
 {
     (void)_netif_add;   /* in case it is not used */
     /* TODO: do for every eligible netdev */
-#ifdef LWIP_NETIF_NUMOF
+#ifdef LWIP_NETIF_USED
```
Comment (miri64) [inline pkg/lwip/contrib/lwip.c:233]: As you define `LWIP_NETIF_NUMOF_...` if the module is not there as 0 and then `LWIP_NETIF_NUMOF` as the sum of those you could instead of the `#undef; #define` mess above just remove `LWIP_NETIF_USED` altogether and use

```suggestion
#if IS_ACTIVE(LWIP_NETIF_NUMOF)
```

or

```suggestion
#if LWIP_NETIF_NUMOF > 0
```

(as it is always defined) here. I think if you use C conditionals here
```suggestion
    if (LWIP_NETIF_NUMOF > 0) { // ...
```

we can even remove most of the pre-processor mess in this file (the array definitions might pose a problem though).

Comment (miri64) [reply pkg/lwip/contrib/lwip.c:233]: > (the array definitions might pose a problem though).

Might be worth trying out putting them as `static` variables in function the function.

File: pkg/lwip/contrib/lwip.c (hunk 1 above)
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:234]: >  Why the _ prefix? It usually is used to denote private scope, but for a function variable basically all is private.

Yeah, of course I know that :wink: It was just to make sure to avoid potential name conflicts.

File: pkg/lwip/contrib/lwip.c (hunk 2 above)
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:233]: First, I tried it in that way `#if LWIP_NETIF_NUMOF > 0` but it does not work for `LWIP_NETIF_NUMOF_*` definitions with `ARRAY_SIZE` definitions. Since the preprocessor doesn't know anything about types, the `sizeof` operator is not allowed in conditionals.

I will try `IS_ACTIVE(LWIP_NETIF_NUMOF)`.

Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:233]: Defining the static variables inside an `if`-block with the `IS_USED` macro should work, for example
```c
if (IS_USED(MODULE_NETDEV_TAP)) {
    static struct netif netif[LWIP_NETIF_NUMOF];
    for (unsigned i = 0; i < LWIP_NETIF_NUMOF_NETDEV_TAP; i++) {
        netdev_tap_setup(&netdev_taps[i], &netdev_tap_params[i]);
        if (_netif_add(&netif[_netif++], &netdev_taps[i], lwip_netdev_init,
                       tcpip_input) == NULL) {
            DEBUG("Could not add netdev_tap device\n");
            return;
        }
    }
}

But would it would produce changes for the whole file because of the indentation. I didn't wanted to change to much with this PR. If you think that it makes sense, I can change it.

Comment (miri64) [reply pkg/lwip/contrib/lwip.c:233]: As this file is not very thought-through to begin with, let's do this now that we make this new step ;-). An alternative, after sleeping on it, could be to split the devices up file-wise like we do with GNRC.

File: pkg/lwip/contrib/lwip.c (hunk 1 above)
Comment (miri64) [reply pkg/lwip/contrib/lwip.c:234]: Then 
```suggestion
    int netif_num = 0;
```
```suggestion
    int netif_count = 0;
```
or something of that kind?

File: pkg/lwip/contrib/lwip.c (hunk 2 above)
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:233]: Unfortunately, we have to use the preprocessor conditional `#if IS_USED(MODULE_NETDEV_TAP)`, with C conditional the code is compiled and `netdev_tap_t` is unknown type :worried:

Comment (miri64) [reply pkg/lwip/contrib/lwip.c:233]: Yes we have to remove the `#ifdef` around the include then of course as well!

File: tests/lwip/main.c [hunk 3]
```diff
@@ -39,7 +39,7 @@ static int ifconfig(int argc, char **argv)
     (void)argc;
     (void)argv;
     for (struct netif *iface = netif_list; iface != NULL; iface = iface->next) {
-        printf("%s_%02u: ", iface->name, iface->num);
+        printf("%c%c_%02u: ", iface->name[0], iface->name[1], iface->num);
```
Comment (miri64) [inline tests/lwip/main.c:42]: Why this change btw?

Comment (gschorcht) [reply tests/lwip/main.c:42]: Because `netif.name` is defined as `char name [2]` and set to `E` + `T` here https://github.com/RIOT-OS/RIOT/blob/ae6e48cd677bfc49df3b6058a3f805ace3069bcf/pkg/lwip/contrib/netdev/lwip_netdev.c#L111-L112 There is no terminating `\0`. So can't be printed as string with `%s`. With `%s` you get garbage if the `netif.num` (the subsequent element) is not 0. It works for netif 0 but not for netifs > 0. That's probably why we didn't realize that.

**Comment by gschorcht** [comment]
@miri64 Do you know whether this would also work in reverse order? https://github.com/RIOT-OS/RIOT/blob/ae6e48cd677bfc49df3b6058a3f805ace3069bcf/pkg/lwip/contrib/lwip.c#L334-L340 I'm asking because`dhcp_start` is only called if mode `netdev_tap` is used and would belong to the block inside the C conditional `if (LWIP_NETIF_NUMOF) { ... }`. Also the static variable `netif` is then only defined inside the C conditional. Since DHCP is using, I don't know how this function behaves if the TCP/IP thread isn't started when it is called.

**Comment by miri64** [comment]
> Since DHCP is using UDP, I don't know how `dhcp_start` function behaves if the TCP/IP thread isn't started when it is called.

The TCP/IP thread has nothing to do with a protocol being UDP or DCCP or whatever, as it is the main event thread of lwIP ;-). I don't remember exactly what happens if you call `dhcp_start()` before the thread is running, but suffice to say, it did not work.

**Comment by gschorcht** [comment]
> @miri64 Do you know whether this would also work in reverse order?

Ok, I know it now, it seems to crash in reverse order.

File: pkg/lwip/contrib/lwip.c (hunk 2 above)
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c]: I pushed a commit that places all static variables inside the function. Wherever possible, it replaces preprocessor conditions with C conditions.

File: pkg/lwip/contrib/lwip.c [hunk 4]
```diff
@@ -179,103 +163,146 @@ void lwip_bootstrap(void)
...
+         * not emit a `NETDEV_EVENT_LINK_UP` event. Remove, once it does
+         * at an appropriate point.
+         * (see https://github.com/RIOT-OS/RIOT/pull/14150) */
+        dhcp_start(netif);
 #endif
+    }
+    else {
+        /* also allow for external interface definition */
+        tcpip_init(NULL, NULL);
+    }
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:305]: This is not really clear to me, if `IS_USED(MODULE_NETDEV_TAP` is true, `LWIP_NETIF_NUMOF` will be greater than 0 anyway.

File: pkg/lwip/contrib/lwip.c (hunk 1 above)
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c]: I renamed it to `netif_idx`.

File: pkg/lwip/contrib/lwip.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (miri64) [reply pkg/lwip/contrib/lwip.c:305]: I just followed your head start and reduced the code duplication. In your code the `dhcp_start(netif)` is in the if conditional for `LWIP_NETIF_NUMOF` as well, but maybe you are right and it should be initialized when `netif != NULL`.

[the remark this answers is not in the document]
Comment (miri64) [reply pkg/lwip/contrib/lwip.c:305]: Ah, since `netif` is an array, `LWIP_NETIF_NUMOF` should be checked after all.

File: pkg/lwip/contrib/lwip.c [hunk 5]
```diff
@@ -179,103 +163,146 @@ void lwip_bootstrap(void)
...
-#ifdef LWIP_NETIF_NUMOF
-#ifdef MODULE_NETDEV_TAP
-    for (unsigned i = 0; i < LWIP_NETIF_NUMOF; i++) {
-        netdev_tap_setup(&netdev_taps[i], &netdev_tap_params[i]);
-        if (_netif_add(&netif[i], &netdev_taps[i], lwip_netdev_init,
-                       tcpip_input) == NULL) {
-            DEBUG("Could not add netdev_tap device\n");
-            return;
+    if (LWIP_NETIF_NUMOF > 0) {
+        static struct netif netif[LWIP_NETIF_NUMOF];
```
Comment (miri64) [inline pkg/lwip/contrib/lwip.c:167]: This does not work for the `lwip_sock` tests :-( (so for externally initialized `netif`s either). Maybe we should make this a global definition after all again and `#ifdef`/`#else` it as a `NULL` pointer if no netifs are defined?

File: pkg/lwip/contrib/lwip.c [hunk 6]
```diff
@@ -23,49 +23,49 @@
 #include "lwip/netif.h"
 #include "netif/lowpan6.h"
 
-#ifdef MODULE_NETDEV_TAP
+#if IS_USED(MODULE_NETDEV_TAP)
```
[the remark this answers is not in the document]
Comment (miri64) [reply pkg/lwip/contrib/lwip.c:26]: Huh? This should only be the case for packages, but as far as I can see, there are no package includes here.

File: pkg/lwip/contrib/lwip.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:305]: I just noticed that `dhcp_start(netif)` only works for the first `netdev_tap` in this way anyway, and only if the `netdev_tap` are added first. I agree that `netifs` would suggest that it refers to multiple netifs.

File: pkg/lwip/contrib/lwip.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:26]: Only if you enable the driver module, `drivers/Makefile.include` adds the drivers include path to `INDLUCES` which is necessary to find `<driver>_params.h`. Furthermore, including `esp_*_netdev.h` works only for ESP boards. I guess it is the same `stm32_eth.h`.

[the remark this answers is not in the document]
Comment (miri64) [reply pkg/lwip/contrib/lwip.c:26]: You are right! :sweat_smile: I'm too used to work only in `sys` and `sys/net`.

[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:26]: Another problem happens if only the `<driver>_params.h` files is guarded and C conditional is used in the function, for example:
```c
        if (IS_USED(MODULE_MRF24J40)) {
            static mrf24j40_t mrf24j40_devs[LWIP_NETIF_NUMOF_MRF24J40];
            for (unsigned i = 0; i < LWIP_NETIF_NUMOF_MRF24J40; i++) {
                mrf24j40_setup(&mrf24j40_devs[i], &mrf24j40_params[i]);
                if (netif_add(&netif[netif_idx++], &mrf24j40_devs[i],
                              lwip_netdev_init, tcpip_6lowpan_input) == NULL) {
                    DEBUG("Could not add mrf24j40 device\n");
                    return;
                }
            }
        }
```
If the `mrf24j40` driver module isn't enabled, this leads to the following 2 errors:
```
ISO C forbids zero-size array ‘mrf24j40_devs’ [-Werror=pedantic]
             static mrf24j40_t mrf24j40_devs[LWIP_NETIF_NUMOF_MRF24J40];
                               ^~~~~~~~~~~~~
```
```
comparison of unsigned expression < 0 is always false [-Werror=type-limits]
             for (unsigned i = 0; i < LWIP_NETIF_NUMOF_MRF24J40; i++) {
                                    ^
```

[the remark this answers is not in the document]
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:26]: Sorry, my page was not in sync, I wrote my last comment before I could see your last comment.

File: pkg/lwip/contrib/lwip.c (hunk 5 above)
Comment (gschorcht) [reply pkg/lwip/contrib/lwip.c:167]: > This does not work for the `lwip_sock` tests :-( (so for externally initialized `netif`s either).

Do you mean the error message error:
```
ISO C forbids zero-size array ‘netif’ [-Werror=pedantic]
         static struct netif netif[LWIP_NETIF_NUMOF];
                             ^~~~~
```
This could be solved with `CFLAGS += -Wno-pedantic` in `pkg/lwip/contrib/Makefile` only for `lwip_contrib` module. With this C flag, `lwip_*` tests are working for me.

Comment (miri64) [reply pkg/lwip/contrib/lwip.c:167]: Yes, but we don't want to deactivate this warning because of that ^^.

**Comment by fjmolinas** [comment]
ping @miri64 @gschorcht :)

File: pkg/lwip/contrib/lwip.c (hunk 5 above)
Comment (miri64) [reply pkg/lwip/contrib/lwip.c:167]: Then maybe the `if` should be an `#if` after all.
